### DISCUSSION TITLE
[Appello fac-simile] Modifica CPU

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#70] [Modifiche architettura MIPS (M-Z)]
![immagine](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/969830c6-18e9-4b27-9d25-addb8d9a0fb8)
![CPU](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/d726fa91-c2e8-4caf-925a-57099545353f)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: 
- [Disegno.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665681/AE.-.Appello.fac-simile.-.E2.-.CPU.pdf)
- [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665690/AE.-.Appello.fac-simile.-.E2.-.Scheda.di.calcolo.pdf)

**Comment by FeddyLix17** [comment]
## 1) Modifiche all'architettura

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/cdd24d8a-806c-4842-a691-a40b75077155)

## 2) Word che esprime l'istruzione *sum_w_mem $a0, $s1, $t2, 0x1D*

- l'Opcode dell'istruzione sum_ w_ mem (fornito in esadecimale dalla traccia) è uguale a $\Large 0x33_{16}  =  00110011_2$ , codificato in 6 bit 110011

- il campo $rs, rappresentate il primo registro sorgente, è $a0, codificato in 5 bit come 00100 (registro numero 4)

- il campo $rt, rappresentante il secondo registro sorgente, è $s1, codificato in 5 bit come 10001 (registro numero 17)

- il campo $rd, rappresentante il registro destinazione, è $t2, codificato in 5 bit come  01010 (registro numero 10)
si ottiene la sequenza completa

- il campo *mem_addr*, tipicamente assegnato al campo *shamt*, è uguale a $\Large 0x1D_{16}  =  00011101_2$, codificato in 5 bit come 11101

- il campo *funct*,  il quale conterrà il codice per l'addizione, è codificato in 6 bit come 100000

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/0fa56a89-c2a2-4f2a-965d-85c474c7aa5a)

## 3) Durata totale ciclo di clock

Considerando

- la fase di *Instruction Fetch*, nella quale si accederà alla memoria, con tempo *180 ns*

- la fase di *Instruction Decode*, nella quale si accederà ai registri, con tempo *60 ns*

- la fase di *Execute*, nella quale verrano parallelamente

    - sommati i contenuti dei registri $rs e $rt
    
    - sommato il valore $\Large 0x1000000_{16}$ a $\Large mem\ addr * 4$

    con tempo *120 ns* 

- la fase di *Memory access*, nella quale si accederà nuovamente alla memoria, con tempo *180 ns*

- una fase nella quale verranno sommati $\Large rs + rt + (4 * mem addr +  0x1000000_{16})$, con tempo *120 ns*

- la fase di *Write Back*, nella quale si accederà nuovamente ai registri per andare a conservare in $rd il risultato della precedente somma, con tempo *60 ns*

la durata totale di un ciclo di clock tale da permettere l'esecuzione della nuova istruzione sarà uguale a

$$
    \Large 60 * 2 + 120 * 2 + 180 * 2 = 720 ns
$$
